### ISSUE TITLE
SSL/TLS support

### ISSUE THREAD

**Issue by avishai-ish-shalom** [riemann/riemann-ruby-client#23]
Initial implementation [here](https://github.com/avishai-ish-shalom/riemann-ruby-client/tree/ssl) - not very happy about it because some openssl specific constants are in `lib/riemann/client/tcp_socket.rb` [here](https://github.com/avishai-ish-shalom/riemann-ruby-client/blob/ssl/lib/riemann/client/tcp_socket.rb#L297) and [here](https://github.com/avishai-ish-shalom/riemann-ruby-client/blob/ssl/lib/riemann/client/tcp_socket.rb#L318)
Another issue is that the current interface of the client uses the same port for UDP and TCP which is inconvenient for TLS (most people bind on port 5554). I'm guessing people who use TLS probably don't want non-encrypted connections and will avoid UDP entirely.

Usage example:
```ruby
c = Riemann::Client.new(ssl: true, key_file: "riemann_server.key", cert_file: "riemann_server.crt", port: 5554) 
c.query("true")
c.tcp << {host: "hlkhlkj", service: "test service", state: "ok", description: "test", metric: 1.222}
```

cc @jamtur01

**Comment by eric** [comment]
I believe that ruby establishes a connection when `TCPSocket.new()` is called.

Some of the changes you made appear to thwart the existing logic that implements timeouts on establishing connections.

Instead of overriding `blank_socket`, you may be better off wrapping `connect_nonblock` with your SSL logic and return the SSL socket from the `socket` method in your `SSLSocket` implementation.

**Comment by avishai-ish-shalom** [comment]
great input, i re-wrote as per your suggestions, much better now. There are still OpenSSL contestants in TcpSocket unfortunately - i could override `readpartial` and `write` but that seems excessive.

**Comment by eric** [comment]
My thought would be that it would make more sense to override the methods instead of including exceptions in a class that isn't relevant.

**Comment by avishai-ish-shalom** [comment]
Makes sense. Another question regarding `Riemman::Client` interface: do we want to have the TLS client exposed under `client.tcp`? the current implementation which automatically chooses between UDP and TCP will probably be useless for people using TLS and `client.tcp` may confuse people.

**Comment by avishai-ish-shalom** [comment]
I added spec tests, however, i'm not sure if I sure save the certificate in fixtures as at least the ca should match riemann instance certificates. How is testing done? i don't see the test config of riemann in the repo

**Comment by jamtur01** [comment]
@avishai-ish-shalom Want to submit this as a PR and we'll do from there?

**Comment by jamtur01** [comment]
Closed in favor of #24.
